subsys: usb: class: hid: Add Out interrupt endpoint

SavinayDharmappa opened the pull request:
Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

nashif commented:
Your commit message needs more details...

finikorg commented on boards/xtensa/intel_s1000_crb/Kconfig.defconfig:
```diff
@@ -73,6 +73,8 @@ config USB_DW_IRQ_PRI
 	def_bool y
+config  USB_DEVICE_HID_BOOTP
+	def_bool y
 endif # USB
```
This does not look good, you do not select HID but try to select HID_BOOTP

finikorg commented on subsys/usb/class/hid/core.c:
```diff
@@ -149,11 +149,19 @@ static void hid_int_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+	SYS_LOG_DBG("ep %x status %d", ep, ep_status);
+}
+
+
```
two empty lines

SavinayDharmappa replied:
done.

finikorg commented on subsys/usb/usb_descriptor.c:
```diff
@@ -622,6 +623,18 @@ static struct dev_common_descriptor common_desc = {
+			.bmAttributes = USB_DC_EP_INTERRUPT,
+			.wMaxPacketSize =
+				sys_cpu_to_le16(
+			CONFIG_HID_INTERRUPT_EP_MPS),
```
In the web it looks like not aligned like previous structure

SavinayDharmappa replied:
I am not sure why it is happening

finikorg commented on subsys/usb/class/hid/core.c:
```diff
@@ -149,11 +149,19 @@ static void hid_int_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+static void hid_int_out(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+{
+	SYS_LOG_DBG("ep %x status %d", ep, ep_status);
+}
```
Missing API and sample

SavinayDharmappa replied:
I have modified it. Please have a look

finikorg commented on subsys/usb/class/hid/core.c:
```diff
@@ -149,11 +149,23 @@ static void hid_int_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+
+	usb_read(ep, data, sizeof(data), &len);
+
+	hid_device.ops->int_out_read(data, len);
```
what if there is no callback?

SavinayDharmappa replied:
ok. I will read the data only if call back is registered othewise  i will just return. Is that ok.

finikorg commented on subsys/usb/class/hid/core.c:
```diff
@@ -149,11 +149,28 @@ static void hid_int_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+static void hid_int_out(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+{
+	u8_t data[CONFIG_HID_INTERRUPT_EP_MPS];
+	int len = 0;
```
Now it is better but we can move stack variable data[] inside {} so that it would not always take memory, something like:

```
...
if (hid_device.ops->int_out_read && ...) {
    u8_t data[CONFIG_HID_INTERRUPT_EP_MPS];
    usb_read(ep, data, sizeof(data), &len);
    ...
}

```

SavinayDharmappa committed:
usb: hid: add a INT OUT transfer complete callback.

This call back is called after interrupt OUT transfer
is completed. This is used send data read from host to
the application

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa committed:
subsys: usb: class: hid: Add OUT interrupt endpoint

patch add a OUT interrupt endpoint descriptor and registers a
corresponding OUT call back with usb driver, which is invoked
when data is sent to device from host.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa committed:
tests: boards: intel_s1000_crb: Read the usb data sent from the host.

patch modifies hid test application to read the data sent from the
host.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

jfischer-phytec-iot commented on subsys/usb/usb_descriptor.c:
```diff
@@ -112,6 +112,7 @@ struct dev_common_descriptor {
 		struct usb_if_descriptor if0;
 		struct usb_hid_descriptor if0_hid;
 		struct usb_ep_descriptor if0_int_ep;
+		struct usb_ep_descriptor if1_int_ep;
```
Both endpoints belong to Interface 0, please rename it to `if0_int_in_ep` and `if0_int_out_ep`. Should not ep_int_out be optional and configurable, as it is not always necessary?

finikorg replied:
Yes, maybe it does make sense to rename CONFIG_HID_OUT_EP_ADDR to CONFIG_HID_INT_OUT_EP_ADDR and CONFIG_HID_INT_EP_ADDR to CONFIG_HID_INT_IN_EP_ADDR

SavinayDharmappa replied:
ok. I will change and make OUT Endpoint optional

sathishkuttan commented on include/usb/class/usb_hid.h:
```diff
@@ -69,6 +70,7 @@ struct hid_ops {
 	 * the next transfer.
 	 */
 	hid_int_ready_callback int_in_ready;
+	hid_out_ready_callback int_out_read;
```
Wouldn't it be better to enable an INT OUT endpoint based on a configuration variable? Since Endpoints may be a premium in devices, whether or not an INT OUT is added to the system should be a choice and a configuration variable would make it easy to make this choice.

sathishkuttan commented on subsys/usb/usb_descriptor.c:
```diff
@@ -591,7 +594,7 @@ static struct dev_common_descriptor common_desc = {
 			.bInterfaceNumber = FIRST_IFACE_HID,
 			.bAlternateSetting = 0,
-			.bNumEndpoints = 1,
+			.bNumEndpoints = 2,
```
Shouldn't this be dependent on `CONFIG_ENABLE_HID_INT_OUT_EP`?

sathishkuttan commented on subsys/usb/usb_descriptor.c:
```diff
@@ -647,7 +664,7 @@ static struct dev_common_descriptor common_desc = {
 			.bInterfaceNumber = FIRST_IFACE_CDC_EEM,
 			.bAlternateSetting = 0,
-			.bNumEndpoints = 2,
+			.bNumEndpoints = NUMOF_ENDPOINTS_HID,
```
Looks like this `NUMOF_ENDPOINTS_HID` assignment is in the wrong place.
Shouldn't this instead be under `if0` of HID?

SavinayDharmappa replied:
Thanks for pointing it out my bad. I will correct it.

SavinayDharmappa committed:
samples: subsys: usb: hid / hid-mouse: rename CONFIG_HID_INT_EP_ADDR.

rename CONFIG_HID_INT_EP_ADDR to CONFIG_HID_INT_IN_EP_ADDR.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa commented:
@jfischer-phytec-iot  I have addressed your comments

sathishkuttan commented on include/usb/class/usb_hid.h:
```diff
@@ -54,6 +54,9 @@ struct usb_hid_descriptor {
 			u8_t **data);
 typedef void (*hid_int_ready_callback)(void);
+#ifdef CONFIG_ENABLE_HID_INT_OUT_EP
+typedef void (*hid_out_ready_callback)(void *data, u32_t len);
```
Please make `hid_out_ready_callback` a notification similar to `hid_int_ready_callback` and expect the application to do the read.

SavinayDharmappa replied:
I have tried that implementation. I was not able receive the correct data sent from host. so I had put read in callback and added application call back.

sathishkuttan replied:
Could it be due to the `hid_ep_data[]` contents being corrupted per my other comment?

SavinayDharmappa replied:
no. I tried with hid_ep_data[] after changing as per your comment. still that issue is present.

SavinayDharmappa commented on subsys/usb/class/hid/core.c:
```diff
@@ -149,11 +149,30 @@ static void hid_int_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
+		u8_t data[CONFIG_HID_INTERRUPT_EP_MPS];
+		u32_t len = 0;
+
+		usb_read(ep, data, sizeof(data), &len);
```
I have tried that implementation. I was not able receive the correct data sent from host. so I had put read in callback and added application call back.

aurel32 replied:
What do you mean by "I was not able receive the correct data"? The data was corrupted? the wrong length?

I have just tried locally and it works perfectly. Here is the patch I used on top of your patchset:

```
--- a/include/usb/class/usb_hid.h
+++ b/include/usb/class/usb_hid.h
@@ -54,9 +54,6 @@ struct usb_hid_descriptor {
 typedef int (*hid_cb_t)(struct usb_setup_packet *setup, s32_t *len,
 			u8_t **data);
 typedef void (*hid_int_ready_callback)(void);
-#ifdef CONFIG_ENABLE_HID_INT_OUT_EP
-typedef void (*hid_out_ready_callback)(void *data, u32_t len);
-#endif
 
 struct hid_ops {
 	hid_cb_t get_report;
@@ -73,7 +70,7 @@ struct hid_ops {
 	 */
 	hid_int_ready_callback int_in_ready;
 #ifdef CONFIG_ENABLE_HID_INT_OUT_EP
-	hid_out_ready_callback int_out_read;
+	hid_int_ready_callback int_out_ready;
 #endif
 };
 
--- a/subsys/usb/class/hid/core.c
+++ b/subsys/usb/class/hid/core.c
@@ -152,15 +152,11 @@ static void hid_int_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
 #ifdef CONFIG_ENABLE_HID_INT_OUT_EP
 static void hid_int_out(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
 {
-	if (hid_device.ops->int_out_read  &&
-		(ep_status == USB_DC_EP_DATA_OUT)) {
-
-		u8_t data[CONFIG_HID_INTERRUPT_EP_MPS];
-		u32_t len = 0;
-
-		usb_read(ep, data, sizeof(data), &len);
-		hid_device.ops->int_out_read(data, len);
+	if (ep_status != USB_DC_EP_DATA_OUT ||
+	    hid_device.ops->int_out_ready == NULL) {
+		return;
 	}
+	hid_device.ops->int_out_ready();
 }
#endif
```


Then, here is the way I read the data in the application:

```
void hid_out()
{
        u8_t data[64];
        u32_t len = 0;

        usb_read(CONFIG_HID_INT_OUT_EP_ADDR, data, sizeof(data), &len);

        /* use the data */
}
```

SavinayDharmappa replied:
data was corrupted. I have also tested. My test target used usb_dw driver.

aurel32 replied:
For me I tested it on the usb_stm32 driver. That said I don't fully get it, doing the read from the callback is just doing it a few instructions after and it's still done in the workqueue thread. It's only done one function call after, could it be that your workqueue stack is too small?

aurel32 replied:
One of the difference of doing a callback function is that the call to usb_read uses the `CONFIG_HID_INT_OUT_EP_ADDR` endpoint instead of the `ep` endpoint passed to hid_int_out. Could it happen that they don't match?

Also you might want to use a different IN and OUT endpoint addresses to check that the DW driver doesn't mix data from both endpoints.

SavinayDharmappa commented on subsys/usb/class/hid/core.c:
```diff
@@ -149,11 +149,30 @@ static void hid_int_in(u8_t ep, enum usb_dc_ep_cb_status_code ep_status)
 		.ep_cb = hid_int_in,
-		.ep_addr = CONFIG_HID_INT_EP_ADDR
+		.ep_addr = CONFIG_HID_INT_IN_EP_ADDR,
+#ifdef CONFIG_ENABLE_HID_INT_OUT_EP
```
ok.

SavinayDharmappa commented:
@jfischer-phytec-iot do u have any further comments ?

jfischer-phytec-iot commented on samples/subsys/usb/hid-mouse/src/main.c:
```diff
@@ -153,7 +153,7 @@ void main(void)
 
 		report[0] = status;
-		usb_write(CONFIG_HID_INT_EP_ADDR, report,
+		usb_write(CONFIG_HID_INT_IN_EP_ADDR, report,
```
It is not very clean,  these sample are broken between the commits. It would be better to make a patch renaming all CONFIG_HID_INT_EP_ADDR in CONFIG_HID_INT_IN_EP_ADDR. Then a patch that introduces CONFIG_HID_INT_OUT_EP_ADDR.
